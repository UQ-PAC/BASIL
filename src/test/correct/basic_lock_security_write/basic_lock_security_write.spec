Globals:
x: int
z: int

L: z -> true, x -> z == 0bv32
Rely: z == old(z) && x == old(x)
Guarantee: old(z) == 0bv32 ==> x == old(x) && z == old(z)

Subroutine: main
Requires: z != 0bv32
Requires: Gamma_R0 == false
