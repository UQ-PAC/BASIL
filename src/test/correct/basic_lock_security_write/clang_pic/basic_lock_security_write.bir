000005ae: program
0000057c: sub __cxa_finalize(__cxa_finalize_result)
000005af: __cxa_finalize_result :: out u32 = low:32[R0]

0000037d:
00000475: R16 := 0x20000
0000047c: R17 := mem[R16 + 8, el]:u64
00000482: R16 := R16 + 8
00000487: call R17 with noreturn

0000057d: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005b0: __do_global_dtors_aux_result :: out u32 = low:32[R0]

0000028b:
0000028f: #3 := R31 - 0x20
00000295: mem := mem with [#3, el]:u64 <- R29
0000029b: mem := mem with [#3 + 8, el]:u64 <- R30
0000029f: R31 := #3
000002a5: R29 := R31
000002ad: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002b2: R19 := 0x20000
000002b9: R0 := pad:64[mem[R19 + 0x30]]
000002bf: when 0:0[R0] goto %000002bd
000005a4: goto %00000346

00000346:
00000349: R0 := 0x1F000
00000350: R0 := mem[R0 + 0xFB8, el]:u64
00000356: when R0 = 0 goto %00000354
000005a5: goto %0000036d

0000036d:
00000370: R0 := 0x20000
00000377: R0 := mem[R0 + 0x28, el]:u64
0000037c: R30 := 0x770
0000037f: call @__cxa_finalize with return %00000354

00000354:
0000035c: R30 := 0x774
0000035e: call @deregister_tm_clones with return %00000360

00000360:
00000363: R0 := 1
0000036b: mem := mem with [R19 + 0x30] <- 7:0[R0]
000005a6: goto %000002bd

000002bd:
000002c7: R19 := mem[R31 + 0x10, el]:u64
000002ce: R29 := mem[R31, el]:u64
000002d3: R30 := mem[R31 + 8, el]:u64
000002d7: R31 := R31 + 0x20
000002dc: call R30 with noreturn

00000581: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005b1: __libc_start_main_main :: in u64 = R0
000005b2: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005b3: __libc_start_main_arg3 :: in out u64 = R2
000005b4: __libc_start_main_auxv :: in out u64 = R3
000005b5: __libc_start_main_result :: out u32 = low:32[R0]

000001e4:
0000045f: R16 := 0x20000
00000466: R17 := mem[R16, el]:u64
0000046c: R16 := R16
00000471: call R17 with noreturn

00000582: sub _fini(_fini_result)
000005b6: _fini_result :: out u32 = low:32[R0]

00000020:
00000026: #0 := R31 - 0x10
0000002c: mem := mem with [#0, el]:u64 <- R29
00000032: mem := mem with [#0 + 8, el]:u64 <- R30
00000036: R31 := #0
0000003c: R29 := R31
00000043: R29 := mem[R31, el]:u64
00000048: R30 := mem[R31 + 8, el]:u64
0000004c: R31 := R31 + 0x10
00000051: call R30 with noreturn

00000583: sub _init(_init_result)
000005b7: _init_result :: out u32 = low:32[R0]

000004cf:
000004d5: #5 := R31 - 0x10
000004db: mem := mem with [#5, el]:u64 <- R29
000004e1: mem := mem with [#5 + 8, el]:u64 <- R30
000004e5: R31 := #5
000004eb: R29 := R31
000004f0: R30 := 0x5F8
000004f2: call @call_weak_fn with return %000004f4

000004f4:
000004f9: R29 := mem[R31, el]:u64
000004fe: R30 := mem[R31 + 8, el]:u64
00000502: R31 := R31 + 0x10
00000507: call R30 with noreturn

00000584: sub _start(_start_result)
000005b8: _start_result :: out u32 = low:32[R0]

000001a5:
000001aa: R29 := 0
000001af: R30 := 0
000001b5: R5 := R0
000001bc: R1 := mem[R31, el]:u64
000001c2: R2 := R31 + 8
000001c8: R6 := R31
000001cd: R0 := 0x1F000
000001d4: R0 := mem[R0 + 0xFD8, el]:u64
000001d9: R3 := 0
000001de: R4 := 0
000001e3: R30 := 0x6B0
000001e6: call @__libc_start_main with return %000001e8

000001e8:
000001eb: R30 := 0x6B4
000001ee: call @abort with return %000005a7

000005a7:
000005a8: call @call_weak_fn with noreturn

00000587: sub abort()


000001ec:
000004a1: R16 := 0x20000
000004a8: R17 := mem[R16 + 0x18, el]:u64
000004ae: R16 := R16 + 0x18
000004b3: call R17 with noreturn

00000588: sub call_weak_fn(call_weak_fn_result)
000005b9: call_weak_fn_result :: out u32 = low:32[R0]

000001f0:
000001f3: R0 := 0x1F000
000001fa: R0 := mem[R0 + 0xFD0, el]:u64
00000200: when R0 = 0 goto %000001fe
000005a9: goto %000003bd

000001fe:
00000206: call R30 with noreturn

000003bd:
000003c0: goto @__gmon_start__

000003be:
0000048b: R16 := 0x20000
00000492: R17 := mem[R16 + 0x10, el]:u64
00000498: R16 := R16 + 0x10
0000049d: call R17 with noreturn

0000058a: sub deregister_tm_clones(deregister_tm_clones_result)
000005ba: deregister_tm_clones_result :: out u32 = low:32[R0]

0000020c:
0000020f: R0 := 0x20000
00000215: R0 := R0 + 0x30
0000021a: R1 := 0x20000
00000220: R1 := R1 + 0x30
00000226: #1 := ~R0
0000022b: #2 := R1 + ~R0
00000231: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
00000237: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
0000023b: ZF := #2 + 1 = 0
0000023f: NF := 63:63[#2 + 1]
00000245: when ZF goto %00000243
000005aa: goto %0000039f

0000039f:
000003a2: R1 := 0x1F000
000003a9: R1 := mem[R1 + 0xFB0, el]:u64
000003ae: when R1 = 0 goto %00000243
000005ab: goto %000003b2

00000243:
0000024b: call R30 with noreturn

000003b2:
000003b6: R16 := R1
000003bb: call R16 with noreturn

0000058d: sub frame_dummy(frame_dummy_result)
000005bb: frame_dummy_result :: out u32 = low:32[R0]

000002e2:
000002e4: call @register_tm_clones with noreturn

0000058e: sub main(main_argc, main_argv, main_result)
000005bc: main_argc :: in u32 = low:32[R0]
000005bd: main_argv :: in out u64 = R1
000005be: main_result :: out u32 = low:32[R0]

000002e6:
000002ea: R31 := R31 - 0x10
000002f2: mem := mem with [R31 + 0xC, el]:u32 <- 31:0[R0]
000002f7: R8 := 0x1F000
000002fe: R8 := mem[R8 + 0xFC0, el]:u64
00000303: R9 := 1
0000030b: mem := mem with [R8, el]:u32 <- 31:0[R9]
00000312: R10 := pad:64[mem[R31 + 0xC, el]:u32]
00000317: R9 := 0x1F000
0000031e: R9 := mem[R9 + 0xFC8, el]:u64
00000326: mem := mem with [R9, el]:u32 <- 31:0[R10]
0000032b: R0 := 0
00000332: mem := mem with [R9, el]:u32 <- 0
00000339: mem := mem with [R8, el]:u32 <- 0
0000033f: R31 := R31 + 0x10
00000344: call R30 with noreturn

0000058f: sub register_tm_clones(register_tm_clones_result)
000005bf: register_tm_clones_result :: out u32 = low:32[R0]

0000024d:
00000250: R0 := 0x20000
00000256: R0 := R0 + 0x30
0000025b: R1 := 0x20000
00000261: R1 := R1 + 0x30
00000268: R1 := R1 + ~R0 + 1
0000026e: R2 := 0.63:63[R1]
00000275: R1 := R2 + (R1 ~>> 3)
0000027b: R1 := extend:64[63:1[R1]]
00000281: when R1 = 0 goto %0000027f
000005ac: goto %00000381

00000381:
00000384: R2 := 0x1F000
0000038b: R2 := mem[R2 + 0xFE0, el]:u64
00000390: when R2 = 0 goto %0000027f
000005ad: goto %00000394

0000027f:
00000287: call R30 with noreturn

00000394:
00000398: R16 := R2
0000039d: call R16 with noreturn
