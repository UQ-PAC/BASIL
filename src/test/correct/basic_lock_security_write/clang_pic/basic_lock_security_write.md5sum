dad7590f0e0ef32d6316805cb44ee332  correct/basic_lock_security_write/clang_pic/a.out
e0b1bd62613d848375ff41d38e451ed1  correct/basic_lock_security_write/clang_pic/basic_lock_security_write.adt
062c49b5e32ef50e70b7b05021fd945b  correct/basic_lock_security_write/clang_pic/basic_lock_security_write.bir
5edffb7ca3ee13b7f17d42f40c59ba4f  correct/basic_lock_security_write/clang_pic/basic_lock_security_write.relf
c691d963ba80827609779a739a4a7cb5  correct/basic_lock_security_write/clang_pic/basic_lock_security_write.gts
