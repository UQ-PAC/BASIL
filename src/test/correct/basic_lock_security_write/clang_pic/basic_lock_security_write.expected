var {:extern} Gamma_R0: bool;
var {:extern} Gamma_R10: bool;
var {:extern} Gamma_R31: bool;
var {:extern} Gamma_R8: bool;
var {:extern} Gamma_R9: bool;
var {:extern} Gamma_mem: [bv64]bool;
var {:extern} Gamma_stack: [bv64]bool;
var {:extern} R0: bv64;
var {:extern} R10: bv64;
var {:extern} R31: bv64;
var {:extern} R8: bv64;
var {:extern} R9: bv64;
var {:extern} mem: [bv64]bv8;
var {:extern} stack: [bv64]bv8;
const {:extern} $x_addr: bv64;
axiom ($x_addr == 69688bv64);
const {:extern} $z_addr: bv64;
axiom ($z_addr == 69684bv64);
function {:extern} L(mem$in: [bv64]bv8, index: bv64) returns (bool) {
  (if (index == $x_addr) then (memory_load32_le(mem$in, $z_addr) == 0bv32) else (if (index == $z_addr) then true else false))
}

function {:extern} {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function {:extern} gamma_load32(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))
}

function {:extern} gamma_load64(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 7bv64)] && (gammaMap[bvadd64(index, 6bv64)] && (gammaMap[bvadd64(index, 5bv64)] && (gammaMap[bvadd64(index, 4bv64)] && (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))))))
}

function {:extern} gamma_store32(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value]
}

function {:extern} memory_load32_le(memory: [bv64]bv8, index: bv64) returns (bv32) {
  (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))
}

function {:extern} memory_load64_le(memory: [bv64]bv8, index: bv64) returns (bv64) {
  (memory[bvadd64(index, 7bv64)] ++ (memory[bvadd64(index, 6bv64)] ++ (memory[bvadd64(index, 5bv64)] ++ (memory[bvadd64(index, 4bv64)] ++ (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))))))
}

function {:extern} memory_store32_le(memory: [bv64]bv8, index: bv64, value: bv32) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]]
}

function {:extern} {:bvbuiltin "zero_extend 32"} zero_extend32_32(bv32) returns (bv64);
procedure {:extern} rely();
  modifies Gamma_mem, mem;
  ensures (forall i: bv64 :: (((mem[i] == old(mem[i])) ==> (Gamma_mem[i] == old(Gamma_mem[i])))));
  ensures ((memory_load32_le(mem, $z_addr) == old(memory_load32_le(mem, $z_addr))) && (memory_load32_le(mem, $x_addr) == old(memory_load32_le(mem, $x_addr))));
  free ensures (memory_load32_le(mem, 1956bv64) == 131073bv32);
  free ensures (memory_load64_le(mem, 69048bv64) == 1872bv64);
  free ensures (memory_load64_le(mem, 69056bv64) == 1792bv64);
  free ensures (memory_load64_le(mem, 69568bv64) == 69684bv64);
  free ensures (memory_load64_le(mem, 69576bv64) == 69688bv64);
  free ensures (memory_load64_le(mem, 69592bv64) == 1876bv64);
  free ensures (memory_load64_le(mem, 69672bv64) == 69672bv64);

procedure {:extern} rely_transitive();
  modifies Gamma_mem, mem;
  ensures ((memory_load32_le(mem, $z_addr) == old(memory_load32_le(mem, $z_addr))) && (memory_load32_le(mem, $x_addr) == old(memory_load32_le(mem, $x_addr))));

implementation {:extern} rely_transitive()
{
  call rely();
  call rely();
}

procedure {:extern} rely_reflexive();

implementation {:extern} rely_reflexive()
{
  assert ((memory_load32_le(mem, $z_addr) == memory_load32_le(mem, $z_addr)) && (memory_load32_le(mem, $x_addr) == memory_load32_le(mem, $x_addr)));
}

procedure {:extern} guarantee_reflexive();
  modifies Gamma_mem, mem;

implementation {:extern} guarantee_reflexive()
{
  assert ((memory_load32_le(mem, $z_addr) == 0bv32) ==> ((memory_load32_le(mem, $x_addr) == memory_load32_le(mem, $x_addr)) && (memory_load32_le(mem, $z_addr) == memory_load32_le(mem, $z_addr))));
}

procedure main_1876();
  modifies Gamma_R0, Gamma_R10, Gamma_R31, Gamma_R8, Gamma_R9, Gamma_mem, Gamma_stack, R0, R10, R31, R8, R9, mem, stack;
  requires (memory_load32_le(mem, $z_addr) != 0bv32);
  requires (Gamma_R0 == false);
  free requires (memory_load64_le(mem, 69664bv64) == 0bv64);
  free requires (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free requires (memory_load32_le(mem, 1956bv64) == 131073bv32);
  free requires (memory_load64_le(mem, 69048bv64) == 1872bv64);
  free requires (memory_load64_le(mem, 69056bv64) == 1792bv64);
  free requires (memory_load64_le(mem, 69568bv64) == 69684bv64);
  free requires (memory_load64_le(mem, 69576bv64) == 69688bv64);
  free requires (memory_load64_le(mem, 69592bv64) == 1876bv64);
  free requires (memory_load64_le(mem, 69672bv64) == 69672bv64);
  free ensures (Gamma_R31 == old(Gamma_R31));
  free ensures (R31 == old(R31));
  free ensures (memory_load32_le(mem, 1956bv64) == 131073bv32);
  free ensures (memory_load64_le(mem, 69048bv64) == 1872bv64);
  free ensures (memory_load64_le(mem, 69056bv64) == 1792bv64);
  free ensures (memory_load64_le(mem, 69568bv64) == 69684bv64);
  free ensures (memory_load64_le(mem, 69576bv64) == 69688bv64);
  free ensures (memory_load64_le(mem, 69592bv64) == 1876bv64);
  free ensures (memory_load64_le(mem, 69672bv64) == 69672bv64);

implementation main_1876()
{
  var Gamma_load18: bool;
  var Gamma_load19: bool;
  var Gamma_load20: bool;
  var Gamma_x_old: bool;
  var load18: bv64;
  var load19: bv32;
  var load20: bv64;
  var x_old: bv32;
  var z_old: bv32;
  lmain:
    assume {:captureState "lmain"} true;
    R31, Gamma_R31 := bvadd64(R31, 18446744073709551600bv64), Gamma_R31;
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), R0[32:0]), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), Gamma_R0);
    assume {:captureState "%000002f2"} true;
    R8, Gamma_R8 := 65536bv64, true;
    call rely();
    load18, Gamma_load18 := memory_load64_le(mem, bvadd64(R8, 4032bv64)), (gamma_load64(Gamma_mem, bvadd64(R8, 4032bv64)) || L(mem, bvadd64(R8, 4032bv64)));
    R8, Gamma_R8 := load18, Gamma_load18;
    R9, Gamma_R9 := 1bv64, true;
    call rely();
    z_old := memory_load32_le(mem, $z_addr);
    x_old := memory_load32_le(mem, $x_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    assert (L(mem, R8) ==> Gamma_R9);
    mem, Gamma_mem := memory_store32_le(mem, R8, R9[32:0]), gamma_store32(Gamma_mem, R8, Gamma_R9);
    assert ((R8 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old == 0bv32) ==> ((memory_load32_le(mem, $x_addr) == x_old) && (memory_load32_le(mem, $z_addr) == z_old)));
    assume {:captureState "%0000030b"} true;
    load19, Gamma_load19 := memory_load32_le(stack, bvadd64(R31, 12bv64)), gamma_load32(Gamma_stack, bvadd64(R31, 12bv64));
    R10, Gamma_R10 := zero_extend32_32(load19), Gamma_load19;
    R9, Gamma_R9 := 65536bv64, true;
    call rely();
    load20, Gamma_load20 := memory_load64_le(mem, bvadd64(R9, 4040bv64)), (gamma_load64(Gamma_mem, bvadd64(R9, 4040bv64)) || L(mem, bvadd64(R9, 4040bv64)));
    R9, Gamma_R9 := load20, Gamma_load20;
    call rely();
    z_old := memory_load32_le(mem, $z_addr);
    x_old := memory_load32_le(mem, $x_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    assert (L(mem, R9) ==> Gamma_R10);
    mem, Gamma_mem := memory_store32_le(mem, R9, R10[32:0]), gamma_store32(Gamma_mem, R9, Gamma_R10);
    assert ((R9 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old == 0bv32) ==> ((memory_load32_le(mem, $x_addr) == x_old) && (memory_load32_le(mem, $z_addr) == z_old)));
    assume {:captureState "%00000326"} true;
    R0, Gamma_R0 := 0bv64, true;
    call rely();
    z_old := memory_load32_le(mem, $z_addr);
    x_old := memory_load32_le(mem, $x_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    assert (L(mem, R9) ==> true);
    mem, Gamma_mem := memory_store32_le(mem, R9, 0bv32), gamma_store32(Gamma_mem, R9, true);
    assert ((R9 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old == 0bv32) ==> ((memory_load32_le(mem, $x_addr) == x_old) && (memory_load32_le(mem, $z_addr) == z_old)));
    assume {:captureState "%00000332"} true;
    call rely();
    z_old := memory_load32_le(mem, $z_addr);
    x_old := memory_load32_le(mem, $x_addr);
    Gamma_x_old := (gamma_load32(Gamma_mem, $x_addr) || L(mem, $x_addr));
    assert (L(mem, R8) ==> true);
    mem, Gamma_mem := memory_store32_le(mem, R8, 0bv32), gamma_store32(Gamma_mem, R8, true);
    assert ((R8 == $z_addr) ==> (L(mem, $x_addr) ==> Gamma_x_old));
    assert ((z_old == 0bv32) ==> ((memory_load32_le(mem, $x_addr) == x_old) && (memory_load32_le(mem, $z_addr) == z_old)));
    assume {:captureState "%00000339"} true;
    R31, Gamma_R31 := bvadd64(R31, 16bv64), Gamma_R31;
    goto main_1876_basil_return;
  main_1876_basil_return:
    assume {:captureState "main_1876_basil_return"} true;
    return;
}

